massa-ledger/src/ledger.rs

```diff
@@ -106,55 +102,78 @@ impl FinalLedger {
 
     /// Gets a snapshot of the ledger to bootstrap other nodes
     ///
     /// TODO: This loads the whole ledger in RAM. Switch to streaming in the future
     pub fn get_bootstrap_state(&self) -> FinalLedgerBootstrapState {
+        // temporary implementation while waiting for streaming
         FinalLedgerBootstrapState {
-            sorted_ledger: self.sorted_ledger.clone(),
+            sorted_ledger: self
+                .sorted_ledger
+                .get_every_address()
+                .iter()
+                .map(|(addr, balance)| {
+                    (
+                        *addr,
+                        LedgerEntry {
+                            parallel_balance: *balance,
+                            bytecode: self
+                                .sorted_ledger
+                                .get_entry(&addr, LedgerSubEntry::Bytecode)
+                                .unwrap_or(Vec::new()),
+                            datastore: self.sorted_ledger.get_datastore_for(&addr),
+                        },
+                    )
+                })
+                .collect(),
         }
     }
 
-    /// Gets a copy of a full ledger entry.
-    ///
-    /// # Returns
-    /// A clone of the whole `LedgerEntry`, or None if not found.
-    ///
-    /// TODO: in the future, never manipulate full ledger entries because their datastore can be huge
-    /// `https://github.com/massalabs/massa/issues/2342`
-    pub fn get_full_entry(&self, addr: &Address) -> Option<LedgerEntry> {
-        self.sorted_ledger.get(addr).cloned()
-    }
-
     /// Gets the parallel balance of a ledger entry
     ///
     /// # Returns
     /// The parallel balance, or None if the ledger entry was not found
     pub fn get_parallel_balance(&self, addr: &Address) -> Option<Amount> {
-        self.sorted_ledger.get(addr).map(|v| v.parallel_balance)
+        // note: can a balance ever have a invalid format?
+        self.sorted_ledger
+            .get_entry(addr, LedgerSubEntry::Balance)
+            .map(|bytes| {
+                Amount::from_raw(
+                    u64::from_varint_bytes(&bytes)
+                        .expect("critical: invalid balance format")
+                        .0,
+                )
+            })
     }
 
     /// Gets a copy of the bytecode of a ledger entry
     ///
     /// # Returns
     /// A copy of the found bytecode, or None if the ledger entry was not found
     pub fn get_bytecode(&self, addr: &Address) -> Option<Vec<u8>> {
-        self.sorted_ledger.get(addr).map(|v| v.bytecode.clone())
+        self.sorted_ledger.get_entry(addr, LedgerSubEntry::Bytecode)
     }
 
     /// Checks if a ledger entry exists
     ///
     /// # Returns
     /// true if it exists, false otherwise.
     pub fn entry_exists(&self, addr: &Address) -> bool {
-        self.sorted_ledger.contains_key(addr)
+        // note: document the "may"
+        self.sorted_ledger
+            .entry_may_exist(addr, LedgerSubEntry::Balance)
```

Comment:
I do not exactly know where the may come from, but from what've I've red (see https://github.com/facebook/rocksdb/issues/4149#issuecomment-405966488) it seems outdated. I used it in the first place to not break the Ledger interface too much, and because it was qualified as more light weight that using `get()`, but I have seen by myself that it is not trustworthy, forgot to add a comment about it.

Reply:
ok let's avoid using it :)

Reply:
Added a comment about it.

Reply:
Actually I just removed it.